File: CSharp/LeetCode/008-StringToInteger(atoi).cs
namespace LeetCode
{
    public class _008_StringToInteger
    {
        public int MyAtoi(string str)
        {
            if (string.IsNullOrWhiteSpace(str)) { return 0; }

            var sign = 1;
            var index = 0;

            while (index < str.Length && str[index] == ' ') { index++; }

            if (str[index] == '-')
            {
                sign = -1;
                index++;
            }
            else if (str[index] == '+')
            {
                index++;
            }

            var positiveOverflowHead = int.MaxValue / 10;
            var positiveOverflowTail = int.MaxValue % 10;

            var result = 0;
            for (; index < str.Length; index++)
            {
                var ch = str[index];
                if (ch < '0' || ch > '9') { break; }

                if (result > positiveOverflowHead ||
                    (result == positiveOverflowHead && ch - '0' > positiveOverflowTail))
                {
                    return sign == 1 ? int.MaxValue : int.MinValue;
                }

                result = result * 10 + (ch - '0');
            }

            return result * sign;
        }
    }
}

File: C++/LeetCode/008-StringToInteger(atoi).cpp
#include "stdafx.h"
#include "008-StringToInteger(atoi).h"

namespace LeetCode
{
	_008_StringToInteger::_008_StringToInteger()
	{
	}


	_008_StringToInteger::~_008_StringToInteger()
	{
	}

	int _008_StringToInteger::myAtoi(string str)
	{
		if (str.empty()) { return 0; }

		bool nagative = false;
		int index = 0;

		while (index < str.size())
		{
			if (!isspace(str[index])) break;
			else index++;
		}

		if (str[index] == '-')
		{
			nagative = true;
			index++;
		}
		else if (str[index] == '+')
		{
			index++;
		}

		int result = 0, digit;
		auto POSITIVE_OVERFLOW_HEAD = INT_MAX / 10;
		auto POSITIVE_OVERFLOW_TAIL = INT_MAX % 10;
		while (index < str.size() && isdigit(str[index]))
		{
			digit = str[index] - '0';
			if (result > POSITIVE_OVERFLOW_HEAD ||
				(result == POSITIVE_OVERFLOW_HEAD && digit > POSITIVE_OVERFLOW_TAIL))
			{
				return nagative ? INT_MIN : INT_MAX;
			}

			result = result * 10 + digit;
			index++;
		}

		return nagative ? -result : result;
	}
}